### ISSUE TITLE
Deluge + Basic Auth cannot connect

### ISSUE THREAD

**Issue by seanfarley** [tympanix/Electorrent#177]
**Client:**
- [ ] µTorrent
- [ ] qBittorrent
- [ ] Transmission
- [ ] rTorrent
- [x] Deluge

**Client Version:**
2.7.1.354

**Operating System:**
MacOS

**Application Version:**
Deluge 2.0.3-2 (linuxserver/deluge docker image)

**Description:**
I can't seem to get any connection to deluge when it's behind a basic auth reverse proxy. I can connect when on the same network and using the raw ip + port.

Any help debugging would be appreciated.

**Comment by tympanix** [comment]
I don't believe we have Deluge v2 support yet, this might introduce some problems. Any debugging can be done in the developer console (Cmd+alt+i). Support for Deluge is handled by https://github.com/tympanix/node-deluge.

**Comment by seanfarley** [comment]
I can try to debug, thanks :-) But how is basic auth handled at all? In the url (i.e. `https://user:pass@example.com`)? Or something else?

**Comment by tympanix** [comment]
I read through the documentation once again. It is available here: https://buildmedia.readthedocs.org/media/pdf/deluge/develop/deluge.pdf. It seems the authentication is performed with RPC calls, hence Deluge has its own authentication mechanism. This is why, when you add basic auth on top, the RPC calls cannot be performed. [This](https://github.com/tympanix/node-deluge/blob/b60695055f5991e06369d04ad1b2f2a798760ea0/index.js#L64) is where that RPC call is performed. Does this mean that you have authentication disabled in Deluge and rely on basic auth from reverse proxy only? In this case we would have to come up with some clever solution to support in (i.e. authentication plugins or similar) which would be an entirely new feature

**Comment by seanfarley** [comment]
Ah, yes, that would explain it. In general, I put everything behind a proxy with basic auth for just another layer of obscurity. But that is orthogonal to the backend? It *should* (untested hypothesis) be as simple as putting `Authorization: Basic ZWx1c3VhcmlvOnlsYWNsYXZl` (the md5 hash of the user name and password) in the request headers.

Though, I don't know how to hook into the request headers at all.

**Comment by tympanix** [comment]
Yes, you're right, that header would most likely do the trick. Now, everybody wouldn't need that header, since only some users will use reverse proxy, which means it would be a plugin option. I'm thinking of some sort of request middleware, which can hook into the API implementation. This would, in general, provide plugins/middleware for every client implementation.

**Comment by seanfarley** [comment]
A plug-in system sounds nice but a bit overkill for a checkbox to enable a separate new text box for username / password 🤷 

Unless there’s already a plug-in system or middleware system I can currently use?

**Comment by tympanix** [comment]
You could implement said checkbox which would solve your immediate problem. It would not solve the problem for other clients which makes it a one-off thing. We can do it quick and dirty, or we can do it generally (i.e. for any reverse proxy authentication method).

There is currently no plugin/middle system though.

**Comment by seanfarley** [comment]
> It would not solve the problem for other clients which makes it a one-off thing

This is the part I don't understand. It would solve it for any basic auth reverse-proxy, though.

> We can do it quick and dirty, or we can do it generally (i.e. for any reverse proxy authentication method).

I guess you mean more than just basic auth? If so, then yes, it would only work for basic auth (unless we enumerate different auth protocols in a dropdown or something). I just want to make sure there isn't any confusion that this is a deluge-only thing (it'd be for any service behind a basic auth protection).

I can try to take a stab at something (and also deluge v2) but I don't think I can take on designing a middleware system, though. What do you think?

**Comment by tympanix** [comment]
Sorry for my late reply. Certainly, if you feel for it, I would be happy to look through a PR with those features.

My thinking was, as you mention, that the implementation should not be specific to Deluge. That's all. Let's postpone a plugin/middleware system for now.

**Comment by tympanix** [comment]
Closing this as we can't easily support arbitrary reverse proxy authentication. Best option here would be to use the auth which comes with Deluge. For external networks maybe use a transparent protocol like VPN.
